pkg/controller/deadmanssnitchintegration/deadmanssnitchintegration_controller.go

```diff
@@ -0,0 +1,530 @@
+var _ reconcile.Reconciler = &ReconcileDeadmansSnitchIntegration{}
+
+// ReconcileDeadmansSnitchIntegration reconciles a DeadmansSnitchIntegration object
+type ReconcileDeadmansSnitchIntegration struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client    client.Client
+	scheme    *runtime.Scheme
+	dmsclient func(apiKey string, collector *localmetrics.MetricsCollector) dmsclient.Client
+}
+
+// Reconcile reads that state of the cluster for a DeadmansSnitchIntegration object and makes changes based on the state read
+// and what is in the DeadmansSnitchIntegration.Spec
+func (r *ReconcileDeadmansSnitchIntegration) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling DeadmansSnitchIntegration")
+	// Fetch the DeadmansSnitchIntegration dmsi
+	dmsi := &deadmanssnitchv1alpha1.DeadmansSnitchIntegration{}
+
+	start := time.Now()
+	defer func() {
+		reconcileDuration := time.Since(start).Seconds()
+		reqLogger.WithValues("Duration", reconcileDuration).Info("Reconcile complete.")
+		localmetrics.Collector.ObserveReconcile(reconcileDuration)
+	}()
+
+	err := r.client.Get(context.TODO(), request.NamespacedName, dmsi)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	dmsAPIKey, err := utils.LoadSecretData(r.client, dmsi.Spec.DmsAPIKeySecretRef.Name,
+		dmsi.Spec.DmsAPIKeySecretRef.Namespace, DeadMansSnitchAPISecretKey)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	dmsc := r.dmsclient(dmsAPIKey, localmetrics.Collector)
+
+	matchingClusterDeployments, err := r.getMatchingClusterDeployment(dmsi)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if dmsi.DeletionTimestamp != nil {
+		for _, clustDeploy := range matchingClusterDeployments.Items {
+			err = r.deleteDMSI(dmsi, &clustDeploy, dmsc)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+		}
+		return reconcile.Result{}, nil
+	}
+
+	for _, clustDeploy := range matchingClusterDeployments.Items {
+		if clustDeploy.DeletionTimestamp != nil || clustDeploy.Labels[config.ClusterDeploymentNoalertsLabel] == "true" {
+			err = r.deleteDMSClusterDeploy(dmsi, &clustDeploy, dmsc)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+			return reconcile.Result{}, nil
+
+		}
+
+		if !clustDeploy.Spec.Installed {
+			// Cluster isn't installed yet, return
+			return reconcile.Result{}, nil
+		}
+
+		err = r.dmsAddFinalizer(dmsi, &clustDeploy)
```

Comment:
I'm torn about this one.  In general, any update to a watched resource should be followed by exiting the reconcile loop. 
 Every time you update a watched resource a new event is triggered which calls Reconcile. That means;

- dmsAddFinalizer
- createSecret
- createSyncSet

But these are all idempotent and small.  Only issue might be attempted create of the secret or syncset in two threads if the operator supports it.  And that would just mean one fails and is requeued and is a noop for the next reconcile.  Thoughts?  I think this is something I didn't consider with the pagerduty-operator updates.

Reply:
@jewzaam we actually _did_ discuss this a little in the paterduty-operator changes :grin: https://github.com/openshift/pagerduty-operator/pull/90#discussion_r448159838

> Only issue might be attempted create of the secret or syncset in two threads if the operator supports it.

In the above linked conversation, I made the case that there won't be two threads reconciling at the same time (unless we deliberately change things to allow such non-default behaviour), but rather than try to argue the point too much (I might well be wrong), I just changed it (but just that one case that was highlighted, it looks like I didn't consider the other cases...I think that was probably because I was thinking "maybe their point is that a completely separate process, such as hive or another operator, might try to reconcile" since it was on adding the finalizer to the ClusterDeployment; but thinking about it now, I don't think that's what you meant): https://github.com/openshift/pagerduty-operator/pull/90/commits/20f2b769e3bfd128edb90233b0e80abe4518077f

I do see that the existing code has a comment that says "strange things will happen" if the reconcile is not exited early after update (there wasn't any such thing in the pagerduty-operator before the new changes there, and the reconcile wasn't exited after updating the finalizer): https://github.com/openshift/deadmanssnitch-operator/blob/master/pkg/controller/deadmanssnitch/deadmanssnitch_controller.go#L157..L160

So, I'm also torn :slightly_smiling_face: While this operator (before these changes) does return early after adding the finalizer to the ClusterDeployment, it doesn't return early after the creating the secret or syncset. I think to do it properly for the create/update (currently just create happens, I think), we'd need to use `controllerutil.CreateOrUpdate` (which is awesome), and return and requeue if the operation result is not None. I think that would be better done in a separate PR after these changes, WDYT?

@R-Lawton I don't think it would be a huge amount of work to change this particular case to return early after adding finalizers to DMSI and CD resources, WDYT?

Reply:
I agree i think it would be a good idea. I think a separate PR also make sense as well

Reply:
Thanks for the reminder of that discussion on the pagerduty-operator!  I was worried I was giving inconsistent feedback.. whew.

That being said, yes I think this is a small thing and yes if we want to change it a separate PR is fine.  The thing that is keeping me torn is that reconcile is against the DMSI, so every OSD will be reconciled on any cluster create, cluster update, cluster delete.  This is such a simple operator and simple resources I don't see it being a big risk but do think we should be mindful of what happens as we scale up.